Mathlib/MeasureTheory/Measure/Haar/Quotient.lean

```diff
@@ -6,159 +6,385 @@ Authors: Alex Kontorovich, Heather Macbeth
+  apply IsHaarMeasure.smul
+  · intro h
+    haveI i' : IsOpenPosMeasure (volume : Measure G) := inferInstance
+    apply IsOpenPosMeasure.open_pos (interior K) (μ := volume) (self := i')
+    · exact isOpen_interior
+    · exact K.interior_nonempty
+    rw [← le_zero_iff,
+      ← fund_dom_s.measure_zero_of_invariant _ (fun g ↦ QuotientGroup.sound _ _ g) h]
+    apply measure_mono
+    refine interior_subset.trans ?_
+    show (K : Set G) ⊆ π ⁻¹' (π '' K)
+    exact subset_preimage_image π K
+  · show volume (π ⁻¹' (π '' K) ∩ s) ≠ ⊤
+    apply ne_of_lt
+    refine lt_of_le_of_lt ?_ finiteCovol.lt_top
+    apply measure_mono
+    exact inter_subset_right _ s
+
+/- Given a normal subgroup `Γ` of a topological group `G` with Haar measure `μ`, which is also
+  right-invariant, and a finite volume fundamental domain `𝓕`, the quotient map to `G ⧸ Γ`,
+  properly normalized, satisfies `QuotientVolumeEqVolumePreimage`. -/
+@[to_additive IsAddFundamentalDomain.AddQuotientVolumeEqVolumePreimage_HaarMeasure "Given a normal
+subgroup `Γ` of an additive topological group `G` with Haar measure `μ`, which is also
+right-invariant, and a finite volume fundamental domain `𝓕`, the quotient map to `G ⧸ Γ`,
+properly normalized, satisfies `AddQuotientVolumeEqVolumePreimage`."]
+theorem IsFundamentalDomain.QuotientVolumeEqVolumePreimage_HaarMeasure {𝓕 : Set G}
+    (h𝓕 : IsFundamentalDomain Γ.op 𝓕) [IsMulLeftInvariant μ] [SigmaFinite μ]
+    {V : Set (G ⧸ Γ)} (hV : (interior V).Nonempty) (meas_V : MeasurableSet V)
+    (hμK : μ V = volume ((π ⁻¹' V) ∩ 𝓕)) (neTopV : μ V ≠ ⊤) :
+    QuotientVolumeEqVolumePreimage μ := by
+  apply IsMulLeftInvariant.quotientVolumeEqVolumePreimage_of_set (fund_dom_s := h𝓕)
+    (meas_V := meas_V)
+  · rw [hμK]
+    intro c_eq_zero
+    apply IsOpenPosMeasure.open_pos (interior (π ⁻¹' V)) (μ := volume)
+    · simp
+    · apply Set.Nonempty.mono (preimage_interior_subset_interior_preimage continuous_coinduced_rng)
+      apply hV.preimage'
+      simp
+    · apply measure_mono_null (h := interior_subset)
+      apply h𝓕.measure_zero_of_invariant (ht := fun g ↦ QuotientGroup.sound _ _ g)
+      exact c_eq_zero
+  · exact hμK
+  · exact neTopV
+
+variable (K : PositiveCompacts (G ⧸ Γ))
+
+/- Given a normal subgroup `Γ` of a topological group `G` with Haar measure `μ`, which is also
+  right-invariant, and a finite volume fundamental domain `𝓕`, the quotient map to `G ⧸ Γ`,
+  properly normalized, satisfies `QuotientVolumeEqVolumePreimage`. -/
+@[to_additive IsAddFundamentalDomain.AddQuotientVolumeEqVolumePreimage_bubHaarMeasure "Given a
+normal subgroup `Γ` of an additive topological group `G` with Haar measure `μ`, which is also
+right-invariant, and a finite volume fundamental domain `𝓕`, the quotient map to `G ⧸ Γ`,
+properly normalized, satisfies `AddQuotientVolumeEqVolumePreimage`."]
+theorem IsFundamentalDomain.QuotientVolumeEqVolumePreimage_bubHaarMeasure {𝓕 : Set G}
```

Comment:
I don't think `bub` should appear in lemma names; presumably you mean `smul`?

Reply:
Done, thanks!